src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer.Common/BackgroundDocumentGenerator.cs

```diff
@@ -136,7 +138,7 @@ private void OnBackgroundCapturedWorkload()
         // Internal for testing
         internal void Enqueue(DocumentSnapshot document)
         {
-            _foregroundDispatcher.AssertForegroundThread();
+            _foregroundDispatcher.AssertSpecializedForegroundOrUIThread();
```

Comment:
Wherever `AssertSpecializedForegroundOrUIThread` is called indicates code that may be executed by either the specialized foreground thread or the UI thread depending on the situation.

I'm not super confident on these asserts (I basically just looked at the callsites and made educated guesses, and made sure no exceptions were thrown when running), so if any of these asserts look off, e.g. supposed to run exclusively on one thread, let me know!